// unplugin-vue-macros/dist/chunk-ZSB2EBGW.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/index.ts


var _unplugincombine = require('unplugin-combine');

// src/core/exclude-macros.ts
function excludeDepOptimize() {
  return {
    name: "vue-macros-exclude-dep-optimize",
    config() {
      return {
        optimizeDeps: {
          exclude: ["unplugin-vue-macros/macros"]
        }
      };
    }
  };
}

// src/core/plugin.ts
function resolvePlugin(unplugin, framework, options) {
  if (!options) return;
  return unplugin[framework](options);
}

// src/index.ts
var _betterdefine = require('@vue-macros/better-define'); var _betterdefine2 = _interopRequireDefault(_betterdefine);
var _booleanprop = require('@vue-macros/boolean-prop'); var _booleanprop2 = _interopRequireDefault(_booleanprop);
var _chaincall = require('@vue-macros/chain-call'); var _chaincall2 = _interopRequireDefault(_chaincall);
var _config = require('@vue-macros/config');
var _defineemit = require('@vue-macros/define-emit'); var _defineemit2 = _interopRequireDefault(_defineemit);
var _definemodels = require('@vue-macros/define-models'); var _definemodels2 = _interopRequireDefault(_definemodels);
var _defineprop = require('@vue-macros/define-prop'); var _defineprop2 = _interopRequireDefault(_defineprop);
var _defineprops = require('@vue-macros/define-props'); var _defineprops2 = _interopRequireDefault(_defineprops);
var _definepropsrefs = require('@vue-macros/define-props-refs'); var _definepropsrefs2 = _interopRequireDefault(_definepropsrefs);
var _definerender = require('@vue-macros/define-render'); var _definerender2 = _interopRequireDefault(_definerender);
var _defineslots = require('@vue-macros/define-slots'); var _defineslots2 = _interopRequireDefault(_defineslots);
var _devtools = require('@vue-macros/devtools');
var _exportexpose = require('@vue-macros/export-expose'); var _exportexpose2 = _interopRequireDefault(_exportexpose);
var _exportprops = require('@vue-macros/export-props'); var _exportprops2 = _interopRequireDefault(_exportprops);
var _exportrender = require('@vue-macros/export-render'); var _exportrender2 = _interopRequireDefault(_exportrender);
var _hoiststatic = require('@vue-macros/hoist-static'); var _hoiststatic2 = _interopRequireDefault(_hoiststatic);
var _jsxdirective = require('@vue-macros/jsx-directive'); var _jsxdirective2 = _interopRequireDefault(_jsxdirective);
var _namedtemplate = require('@vue-macros/named-template'); var _namedtemplate2 = _interopRequireDefault(_namedtemplate);
var _reactivitytransform = require('@vue-macros/reactivity-transform'); var _reactivitytransform2 = _interopRequireDefault(_reactivitytransform);
var _scriptlang = require('@vue-macros/script-lang'); var _scriptlang2 = _interopRequireDefault(_scriptlang);
var _setupblock = require('@vue-macros/setup-block'); var _setupblock2 = _interopRequireDefault(_setupblock);
var _setupcomponent = require('@vue-macros/setup-component'); var _setupcomponent2 = _interopRequireDefault(_setupcomponent);
var _setupsfc = require('@vue-macros/setup-sfc'); var _setupsfc2 = _interopRequireDefault(_setupsfc);
var _shortbind = require('@vue-macros/short-bind'); var _shortbind2 = _interopRequireDefault(_shortbind);
var _shortemits = require('@vue-macros/short-emits'); var _shortemits2 = _interopRequireDefault(_shortemits);
var _shortvmodel = require('@vue-macros/short-vmodel'); var _shortvmodel2 = _interopRequireDefault(_shortvmodel);
var _unpluginvuedefineoptions = require('unplugin-vue-define-options'); var _unpluginvuedefineoptions2 = _interopRequireDefault(_unpluginvuedefineoptions);

var name = "unplugin-vue-macros";
var plugin = _unplugincombine.createCombinePlugin.call(void 0, (userOptions = {}, meta) => {
  const options = _config.resolveOptions.call(void 0, userOptions);
  const framework = meta.framework;
  const setupComponentPlugins = resolvePlugin(
    _setupcomponent2.default,
    framework,
    options.setupComponent
  );
  const namedTemplatePlugins = resolvePlugin(
    _namedtemplate2.default,
    framework,
    options.namedTemplate
  );
  const plugins = [
    resolvePlugin(_setupsfc2.default, framework, options.setupSFC),
    _optionalChain([setupComponentPlugins, 'optionalAccess', _ => _[0]]),
    resolvePlugin(_setupblock2.default, framework, options.setupBlock),
    resolvePlugin(_scriptlang2.default, framework, options.scriptLang),
    options.plugins.vueRouter,
    _optionalChain([namedTemplatePlugins, 'optionalAccess', _2 => _2[0]]),
    // props
    resolvePlugin(_chaincall2.default, framework, options.chainCall),
    resolvePlugin(_defineprops2.default, framework, options.defineProps),
    resolvePlugin(_definepropsrefs2.default, framework, options.definePropsRefs),
    resolvePlugin(_exportprops2.default, framework, options.exportProps),
    // emits
    resolvePlugin(_defineemit2.default, framework, options.defineEmit),
    resolvePlugin(_shortemits2.default, framework, options.shortEmits),
    // both props & emits
    resolvePlugin(_definemodels2.default, framework, options.defineModels),
    // convert to runtime props & emits
    resolvePlugin(_betterdefine2.default, framework, options.betterDefine),
    // runtime props
    resolvePlugin(_defineprop2.default, framework, options.defineProp),
    resolvePlugin(_defineslots2.default, framework, options.defineSlots),
    resolvePlugin(_exportrender2.default, framework, options.exportRender),
    resolvePlugin(_exportexpose2.default, framework, options.exportExpose),
    resolvePlugin(
      _reactivitytransform2.default,
      framework,
      options.reactivityTransform
    ),
    resolvePlugin(_hoiststatic2.default, framework, options.hoistStatic),
    resolvePlugin(_unpluginvuedefineoptions2.default, framework, options.defineOptions),
    ...framework === "vite" || framework === "rollup" || framework === "rolldown" ? [
      resolvePlugin(
        // VueBooleanProp is not an unplugin, by now
        _booleanprop2.default,
        framework,
        options.booleanProp
      ),
      resolvePlugin(
        // VueShortBind is not an unplugin, by now
        _shortbind2.default,
        framework,
        options.shortBind
      ),
      resolvePlugin(
        // VueShortVmodel is not an unplugin, by now
        _shortvmodel2.default,
        framework,
        options.shortVmodel
      )
    ] : [],
    options.plugins.vue,
    resolvePlugin(_jsxdirective2.default, framework, options.jsxDirective),
    options.plugins.vueJsx,
    resolvePlugin(_definerender2.default, framework, options.defineRender),
    _optionalChain([setupComponentPlugins, 'optionalAccess', _3 => _3[1]]),
    _optionalChain([namedTemplatePlugins, 'optionalAccess', _4 => _4[1]]),
    framework === "vite" ? _devtools.Devtools.call(void 0, { nuxtContext: options.nuxtContext }) : void 0,
    framework === "vite" ? excludeDepOptimize() : void 0
  ].filter(Boolean);
  return {
    name,
    plugins
  };
});
var src_default = plugin;





exports.src_default = src_default; exports.defineConfig = _config.defineConfig; exports.resolveOptions = _config.resolveOptions;
